run.c

```diff
@@ -194,7 +194,10 @@ void softmax(float* x, int size) {
 void matmul(float* xout, float* x, float* w, int n, int d) {
     // W (d,n) @ x (n,) -> xout (d,)
     // by far the most amount of time is spent inside this little function
+    #pragma omp target data map(tofrom: w[0:d*n],x[0:n], xout[0:d])
```

Comment:
I am not an expert when it comes to CUDA but right now this is pretty basic as just mapping the data to GPU and using target directive is not going to utilize the GPU to the full extent. It would be interesting to see a benchmark as it might even be slower than the CPU version. Instead, to make this more optimal, `#pragma omp distribute` and `#pragma omp teams` should be used. What these will do is first create team of threads on the blocks of the GPU with `#pragma omp teams`. Then, the loop iterations will be distributed to the threads of the GPU with `#pragma omp distribute`. An example might look like this. Here, two loops are collapsed into 1, distributed to team of threads on the GPU.

```C
#pragma omp target teams distribute \
          parallel for collapse(2) map(to: x[0:n], w[0:d*n]) map(from: xout[0:d])
for (int i = 0; i < d; i++) { 
    for (int j = 0; j < n; j++) { 
          xout[i] += w[i * n + j] * x[j]; 
     } 
}
```

A good reference/guide from NVIDIA [OpenMP CUDA](https://on-demand.gputechconf.com/gtc/2018/presentation/s8344-openmp-on-gpus-first-experiences-and-best-practices.pdf)

Reply:
Not an expert. Just taking a look at the code. I am not sure if the compiler optimizes computations in the nested for loops, like, e.g.,
for (int i = 0; i < d; i++) { 
   int in = i*n;
    for (int j = 0; j < n; j++) { 
          xout[i] += w[in + j] * x[j]; 
     } 
}

And if readability could be traded to reduce some computation, pointer arithmetic could be used
to access array elements instead of i and j;